### PR TITLE
initial logic for backend

### PR DESCRIPTION
**Pull request by muhdmud**

### CODE REVIEW

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 1]
```diff
@@ -1,14 +1,67 @@
...
+    // id, timestamp, price, size, side
+    arrayData : Array<[id :string, timestamp :number, price : number, size : number, side : string]>;
+    sortedPriceFills : Array<any>;
+    allTimeHigh : number;
+    allTimeLow : number;
+};
+
+const CONNECTION = new Connection('https://solana-api.projectserum.com');
+// mainnet program id
+const PROGRAM_ADDRESS = new PublicKey("9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin");
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:19]: SERUM_V3_PROGRAM_ADDRESS

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 2]
```diff
@@ -1,14 +1,67 @@
...
-export default function handler(
+export default async (
     req: NextApiRequest,
     res: NextApiResponse<Data>
-) {
-    const marketId = req.query.marketId as string
-    res.status(200).json({ name: marketId })
-    // TODO: this should return data on the specific marketId
+) => {
+    const MARKET_ID = req.query.marketId as string
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:25]: marketId

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 3]
```diff
@@ -1,14 +1,67 @@
...
+    //const MARKET_ID_TEST = "8vycGWEqhvnTJR4aTcN2TvtUUNtLXoCTrpCGtoD8R4bo"; 
+    
+    let marketAddress = new PublicKey(MARKET_ID);
+    // check if marketid doesnt exist on mainnet
+    const ifMarketExists = await CONNECTION.getAccountInfo(marketAddress);
+    if (ifMarketExists === null) {
+      throw new Error('Could not find market');
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:32]: return a 400 and return instead

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 4]
```diff
@@ -1,14 +1,67 @@
...
-    res.status(200).json({ name: marketId })
-    // TODO: this should return data on the specific marketId
+) => {
+    const MARKET_ID = req.query.marketId as string
+    //const MARKET_ID_TEST = "8vycGWEqhvnTJR4aTcN2TvtUUNtLXoCTrpCGtoD8R4bo"; 
+    
+    let marketAddress = new PublicKey(MARKET_ID);
+    // check if marketid doesnt exist on mainnet
+    const ifMarketExists = await CONNECTION.getAccountInfo(marketAddress);
+    if (ifMarketExists === null) {
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:31]: try a not ! here

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 5]
```diff
@@ -1,14 +1,67 @@
...
+        totalVolume : 0,
+        arrayData : [],  
+        sortedPriceFills : [],
+        allTimeHigh : 0,
+        allTimeLow : 0,
+    };
+    
+
+    // Retrieving fills
+    for (let fill of await market.loadFills(CONNECTION)) {
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:50]: use .maps and .filters like we discussed in the call

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 6]
```diff
@@ -1,14 +1,50 @@
...
+        totalCalculatedVolume+= f.price;
+        sortedPrices.push(f.price);
+        return { id : f.orderId, price : f.price, size : f.size}
+    });
+
+    sortedPrices.sort();
+    res.status(200).json(  { marketid: marketId, totalFillSize : totalCalculatedFillSize,
+    totalVolume : totalCalculatedVolume, aggregatedData : marketData, sortedPriceFills : sortedPrices,
+    allTimeHigh : highest, allTimeLow : lowest } );
+}
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:50]: add endline

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 7]
```diff
@@ -1,14 +1,50 @@
...
+            highest = f.price;
+        totalCalculatedFillSize++;
+        totalCalculatedVolume+= f.price;
+        sortedPrices.push(f.price);
+        return { id : f.orderId, price : f.price, size : f.size}
+    });
+
+    sortedPrices.sort();
+    res.status(200).json(  { marketid: marketId, totalFillSize : totalCalculatedFillSize,
+    totalVolume : totalCalculatedVolume, aggregatedData : marketData, sortedPriceFills : sortedPrices,
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:48]: aggregatedData -> recentFills

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 8]
```diff
@@ -1,14 +1,50 @@
...
-) {
+    res: NextApiResponse
+) => {
     const marketId = req.query.marketId as string
-    res.status(200).json({ name: marketId })
-    // TODO: this should return data on the specific marketId
-}
+
+    let marketAddress = new PublicKey(marketId);
+    // check if marketid doesnt exist on mainnet
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:18]: uncomment if you wanna keep this

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 9]
```diff
@@ -1,14 +1,60 @@
...
+            'size': f.size,
+            'timestamp': fakeDate.valueOf()
+        }
+        return myData
+    })
+    totalCalculatedVolume = totalSize / marketData.length
+    sortedPrices.sort()
+    let currUniqueHolders = randomIntFromInterval(totalCalculatedVolume, totalCalculatedVolume*2)
+    res.status(200).json(  { marketid: marketId, totalFillSize : marketData.length,
+    totalVolume : totalCalculatedVolume, recentFills : marketData, sortedPriceFills : sortedPrices,
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:57]: stretch this json object out for readability:
```
{
a: 1,
b: 2,
c: 3,
}
```

Comment (muhdmud) [reply pages/api/data/staratlas/markets/[marketId].ts]: done

File: next.config.js [hunk 10]
```diff
@@ -1,4 +1,8 @@
 /** @type {import('next').NextConfig} */
 module.exports = {
   reactStrictMode: true,
+  webpack: (config) => {
```
Comment (dhyon) [inline next.config.js:4]: is this change still necessary?

File: pages/api/data/staratlas/markets.ts [hunk 11]
```diff
@@ -45,4 +45,5 @@ function applyFilter(catalog: Array<any>, queryParams: any) {
     if (queryParams.category)
         result = result.filter(item => item.attributes.category === queryParams.category)
     return result
+
```
Comment (dhyon) [inline pages/api/data/staratlas/markets.ts:48]: please undo

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 12]
```diff
@@ -1,14 +1,68 @@
...
+            'timestamp': fakeDate.valueOf()
+        }
+        return myData
+    })
+    totalCalculatedVolume = totalSize / marketData.length
+    sortedPrices.sort()
+    let currUniqueHolders = randomIntFromInterval(totalCalculatedVolume, totalSize)
+    res.status(200).json(
+    { 
+    marketid: marketId,
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:58]: please indent these fields 1 tab to the right

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 13]
```diff
@@ -1,14 +1,68 @@
...
     req: NextApiRequest,
-    res: NextApiResponse<Data>
-) {
+    res: NextApiResponse
+) => {
     const marketId = req.query.marketId as string
-    res.status(200).json({ name: marketId })
-    // TODO: this should return data on the specific marketId
+
+    let marketAddress = new PublicKey(marketId)
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:22]: const

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 14]
```diff
@@ -1,14 +1,68 @@
...
+    let market = await Market.load(CONNECTION, marketAddress, {}, SERUM_V3_PROGRAM_ADDRESS)
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:23]: const

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 15]
```diff
@@ -1,14 +1,68 @@
...
+    const fills = await market.loadFills(CONNECTION)
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:24]: I think you used to have error handling here? if this fails we need to return a 400 with an error message or we'll just explode with a 500 later down this function

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 16]
```diff
@@ -1,14 +1,68 @@
...
+
+    let lowest = Number.MAX_SAFE_INTEGER
+    let highest = 0
+    let totalCalculatedVolume = 0
+    let totalSize = 0
+    var sortedPrices:string[] = []
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:30]: let

File: pages/api/data/staratlas/markets/[marketId].ts [hunk 17]
```diff
@@ -1,14 +1,68 @@
...
+    let lowest = Number.MAX_SAFE_INTEGER
+    let highest = 0
+    let totalCalculatedVolume = 0
+    let totalSize = 0
+    var sortedPrices:string[] = [] 
+    let fakeDate = new Date()
+    
+
+    const marketData = fills
+    .filter(f => f.side === 'buy')
```
Comment (dhyon) [inline pages/api/data/staratlas/markets/[marketId].ts:35]: when you chain operations on the next line, indent 1 tab to the right

**Review by dhyon — COMMENTED** [review]
just a few style/semantics things

File: next.config.js (hunk 10 above)
Comment (muhdmud) [reply next.config.js:4]: yea no need, since we dont have outer awaits.

File: pages/api/data/staratlas/markets/[marketId].ts (hunk 15 above)
Comment (muhdmud) [reply pages/api/data/staratlas/markets/[marketId].ts:28]: yeah, was testing and missed it. I have re-added it and testing is good!

**Review by dhyon — APPROVED** [review]
great job Mud 👍
